### PR TITLE
Implement Disqus

### PR DESCRIPTION
**Pull request by bradystroud**
Implement Disqus comments
Remove Utterances comments

This PR is not to be merged until we upgrade the SSW Disqus account so we can disable ads

### CODE REVIEW

File: src/components/comments/comments.js [hunk 1]
```diff
@@ -1,34 +1,29 @@
...
+  let disqusConfig = {
+    identifier: props.ruleGuid,
+    title: props.title,
+  };
+  return (
+    <div
+      style={{
+        marginBottom: '5rem',
+        padding: '1rem 1.5rem',
+      }}
```
Comment (pierssinclairssw) [inline src/components/comments/comments.js:15]: Should we be hardcoding this or should it be css?

File: src/pages/profile.js [hunk 2]
```diff
@@ -6,20 +7,43 @@ import ProfileBadge from '../components/profile-badge/profile-badge';
...
   const [dislikedRulesCount, setDislikedRulesCount] = useState();
   const [superDislikedRulesCount, setSuperDislikedRulesCount] = useState();
   const [commentedRulesCount, setCommentedRulesCount] = useState();
 
-  useEffect(() => {}, [listChange, user]);
+  async function CheckUser() {
+    const jwt = await getIdTokenClaims();
+    GetUser(user.sub, jwt.__raw)
+      .then((success) => {
+        setCommentsConnected(success.commentsConnected);
```
Comment (pierssinclairssw) [inline src/pages/profile.js:35]: Does this need to check for the message returned when the user was already connected?

Also does it need to check for response types and process accordingly (e.g. conflict)?

File: src/pages/profile.js [hunk 3]
```diff
@@ -6,20 +7,43 @@ import ProfileBadge from '../components/profile-badge/profile-badge';
...
+      })
+      .catch((error) => {
+        console.error(error);
+      });
```
Comment (pierssinclairssw) [inline src/pages/profile.js:39]: This seems very odd. Is this an existing standard? I would think we don't want to be logging our errors to the console because then anyone can see them if they open the debugger console. Also if there is an error, shouldn't we display some kind of error message or something?

File: src/pages/profile.js [hunk 4]
```diff
@@ -37,10 +61,25 @@ const Profile = ({ data, gitHubUsername }) => {
                 <a
                   className="github-link"
                   href={`https://www.github.com/${user.nickname}`}
+                  target="_blank"
+                  rel="noreferrer"
                 >
                   <GitHubIcon className="profile-github-icon" />
-                  GitHub account | @{user ? user.nickname : ''}
```
Comment (pierssinclairssw) [inline src/pages/profile.js:42]: If the nickname is empty, should we use their user name, full name or something similar as a fallback?

File: src/pages/profile.js [hunk 5]
```diff
@@ -37,10 +61,25 @@ const Profile = ({ data, gitHubUsername }) => {
...
+                    className="disqus-link"
+                    target="_blank"
+                    rel="noreferrer"
+                    href="https://disqus.com/home"
+                  >
+                    <DisqusIcon className="profile-disqus-icon" />
+                    Disqus account
+                  </a>
+                ) : (
+                  <></>
```
Comment (pierssinclairssw) [inline src/pages/profile.js:81]: Can you return nothing here instead of empty brackets or do we need the empty brackets for some reason?

File: src/services/apiService.js [hunk 6]
```diff
@@ -175,15 +176,72 @@ export async function GetSecretContent(Id, token) {
...
+      Authorization: `Bearer ${token}`,
     },
   });
+  return response.json();
+}
 
-  return await response.json();
+export async function ConnectUserCommentsAccount(data, token) {
+  const isEmpty = Object.values(data).some((x) => !x);
+  if (!data || isEmpty) {
```
Comment (pierssinclairssw) [inline src/services/apiService.js:193]: put the isEmpty check in the if statement. I don't think it makes sense to use data in a statement prior to checking if it is defined to begin with

File: src/services/apiService.js [hunk 7]
```diff
@@ -175,15 +176,72 @@ export async function GetSecretContent(Id, token) {
...
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${token}`,
+    },
+    body: JSON.stringify(data),
+  });
+}
+
+export async function RemoveUserCommentsAccount(data, token) {
+  const isEmpty = Object.values(data).some((x) => !x);
+  if (!data || isEmpty) {
```
Comment (pierssinclairssw) [inline src/services/apiService.js:211]: put the isEmpty check in the if statement. I don't think it makes sense to use data in a statement prior to checking if it is defined to begin with

**Review by pierssinclairssw — CHANGES_REQUESTED** [review]
See comments

File: src/pages/profile.js (hunk 2 above)
Comment (bradystroud) [reply src/pages/profile.js:42]: I converted all the error logging to AppInsights

And it will should the user an error message if their is a conflict

**Review by pierssinclairssw — CHANGES_REQUESTED** [review]
See comments

File: src/components/profile-content/profile-content.js [hunk 8]
```diff
@@ -79,33 +99,67 @@ const ProfileContent = (props) => {
...
+        setCommentedRulesList([]);
+      } else {
+        GetDisqusUserCommentsList(success.user.commentsUserId)
+          .then((success) => {
+            if (success.code == 12) {
+              setDisqusPrivacyEnabled(true);
+            } else {
+              const allRules = props.data.allMarkdownRemark.nodes;
+              success.response.forEach((comment) => {
+                if (comment.forum == process.env.DISQUS_FORUM) {
```
[the remark this answers is not in the document]
Comment (bradystroud) [reply src/components/profile-content/profile-content.js:124]: I can't find a way to filter by forum and user 
https://disqus.com/api/docs/

File: src/components/profile-content/profile-content.js [hunk 9]
```diff
@@ -285,37 +346,180 @@ ProfileContent.propTypes = {
...
+                        : 'username-input-box-error'
+                    }
+                    type="text"
+                    name="disqusId"
+                    value={disqusUsername}
+                    placeholder="Enter Disqus Username"
+                    onChange={(e) => setDisqusUsername(e.target.value)}
+                  />
+                  <div className="forgot-username-tooltip">
+                    {/* eslint-disable-next-line jsx-a11y/anchor-has-content */}
```
[the remark this answers is not in the document]
Comment (bradystroud) [reply src/components/profile-content/profile-content.js:407]: We use pseudo elements for some icons so the <a> tag is empty.  When you have an empty <a> tag, eslint puts a red line under it. The comment just disables the red line.

File: src/components/profile-content/profile-content.js [hunk 10]
```diff
@@ -79,33 +100,81 @@ const ProfileContent = (props) => {
...
+  async function createGuidList(response) {
+    var guids = response.map((comment) => {
+      return getGuidFromDisqus(comment);
+    });
+
+    return await Promise.all(guids);
+  }
+
+  async function setCommentedRulesFromGuids(reponse) {
+    const commentedRuleGuids = await createGuidList(reponse);
```
Comment (pierssinclairssw) [inline src/components/profile-content/profile-content.js:136]: This section is concerning because it is going to hit the disqus api for every comment the user has made. I think we can do it better by calling list posts for the user with the related thread argument. We can then use the identifiers on the related thread to identify the associated rule.

**Review by pierssinclairssw — CHANGES_REQUESTED** [review]
See comments

File: src/components/profile-content/profile-content.js [hunk 11]
```diff
@@ -106,50 +107,64 @@ const ProfileContent = (props) => {
       });
   }
 
+  async function getGuidFromDisqus(comment) {
```
[the remark this answers is not in the document]
Comment (bradystroud) [reply src/components/profile-content/profile-content.js:110]: This function is no longer needed and was removed in the latest commit.

File: src/components/profile-content/profile-content.js [hunk 12]
```diff
@@ -106,50 +107,64 @@ const ProfileContent = (props) => {
...
+            error: new Error(err),
+            severityLevel: 3,
+          });
+        });
+
+      return guid;
+    }
+  }
+
+  async function createGuidList(response) {
```
[the remark this answers is not in the document]
Comment (bradystroud) [reply src/components/profile-content/profile-content.js:127]: This function is no longer needed and was removed in the latest commit.

File: src/components/profile-content/profile-content.js (hunk 10 above)
Comment (bradystroud) [reply src/components/profile-content/profile-content.js]: Thanks for the tip, I made this change in this commit 5664366499c0510fee388e2036cd93d18d203dc9

File: src/components/profile-content/profile-content.js [hunk 13]
```diff
@@ -106,50 +107,64 @@ const ProfileContent = (props) => {
...
-    const commentedRuleGuids = [];
     GetUser(user.sub, jwt.__raw).then((success) => {
       setUserCommentsConnected(success.commentsConnected);
       if (!success.commentsConnected) {
         setCommentedRulesList([]);
       } else {
         GetDisqusUserCommentsList(success.user.commentsUserId)
           .then((success) => {
-            if (success.code == 12) {
+            if (success.code == DisqusError.AccessToLow) {
```
[the remark this answers is not in the document]
Comment (bradystroud) [reply src/components/profile-content/profile-content.js]: Fixed in this commit https://github.com/SSWConsulting/SSW.Rules/pull/468/commits/6280f461c243c1a8677fd2d70a74547e0eb006d6

File: src/components/comments-not-connected/comments-not-connected.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            <span className="tooltiptext">Forgot username? </span>
+          </div>
+          <p className="error-text">{errorMessage}</p>
+        </div>
+
+        <button
+          className="connect-acc-button"
+          onClick={() => {
+            if (disqusUsername) {
+              GetDisqusUser(disqusUsername)
```
Comment (pierssinclairssw) [inline src/components/comments-not-connected/comments-not-connected.js:62]: I would always put javascript logic into a separate area so it is clearly separated from the jsx/html code, but you could argue this is a stylistic choice. What do you think?

**Review by pierssinclairssw — COMMENTED** [review]
See my one comment

File: src/components/comments-not-connected/comments-not-connected.js (hunk 14 above)
Comment (pierssinclairssw) [reply src/components/comments-not-connected/comments-not-connected.js:62]: To clarify I mean making this into a function.

Comment (bradystroud) [reply src/components/comments-not-connected/comments-not-connected.js]: As per our conversation, I agree that the logic should be seperate from the html. I made this change in this commit https://github.com/SSWConsulting/SSW.Rules/pull/468/commits/352601d604a11c40b8859d88e883bc6eb32d2d75

**Review by pierssinclairssw — DISMISSED** [review]
Looks good.

File: src/components/profile-content/profile-content.js [hunk 15]
```diff
@@ -285,37 +341,86 @@ ProfileContent.propTypes = {
...
+                  ? 'Nothing here yet, show us the rules you love!'
+                  : type == 'agree'
+                  ? `There should be more here, don${"'"}t you agree?`
+                  : type == 'disagree'
+                  ? 'What a happy chap you are, no disagreements here!'
+                  : `Let us know what you don${"'"}t like, we are here to improve!`}
+              </p>
+            </div>
+          </>
+        )
```
Comment (pierssinclairssw) [inline src/components/profile-content/profile-content.js:416]: Can we make this more readable? It is hard to follow

**Review by pierssinclairssw — APPROVED** [review]

**Review by pierssinclairssw — APPROVED** [review]
